playback/src/local_file.rs

```diff
@@ -0,0 +1,182 @@
+    let src = File::open(audio_path)?;
+    let mss = MediaSourceStream::new(Box::new(src), Default::default());
+
+    let mut hint = Hint::new();
+    hint.with_extension(extension);
+
+    let meta_opts: MetadataOptions = Default::default();
+    let fmt_opts: FormatOptions = Default::default();
+
+    let mut probed = symphonia::default::get_probe()
+        .format(&hint, mss, &fmt_opts, &meta_opts)
+        .map_err(|_| Error::internal("Failed to probe file"))?;
+
+    let mut artist: Option<String> = None;
+    let mut album_title: Option<String> = None;
+    let mut track_title: Option<String> = None;
+
+    let tags = get_tags(&mut probed).ok_or(Error::internal("Failed to probe audio tags"))?;
+
+    for tag in tags {
+        if let Some(std_key) = tag.std_key {
+            match std_key {
+                StandardTagKey::Album => {
+                    album_title.replace(tag.value.to_string());
+                }
+                StandardTagKey::Artist => {
+                    artist.replace(tag.value.to_string());
+                }
+                StandardTagKey::TrackTitle => {
+                    track_title.replace(tag.value.to_string());
+                }
+                _ => {
+                    continue;
+                }
+            }
+        }
+    }
+
+    let first_track = probed
+        .format
+        .default_track()
+        .ok_or(Error::internal("Failed to find an audio track"))?;
+
+    let time_base = first_track
+        .codec_params
+        .time_base
+        .ok_or(Error::internal("Failed to calculate track duration"))?;
+
+    let num_frames = first_track
+        .codec_params
+        .n_frames
+        .ok_or(Error::internal("Failed to calculate track duration"))?;
+
+    let time = time_base.calc_time(num_frames);
+
+    Ok(SpotifyUri::Local {
+        artist: format_uri_component(artist),
+        album_title: format_uri_component(album_title),
+        track_title: format_uri_component(track_title),
+        duration: Duration::from_secs(time.seconds),
+    })
+}
+
+fn get_tags(probed: &mut ProbeResult) -> Option<Vec<Tag>> {
+    if let Some(metadata_rev) = probed.format.metadata().current() {
+        return Some(metadata_rev.tags().to_vec());
+    }
+
+    if let Some(metadata_rev) = probed.metadata.get().as_ref().and_then(|m| m.current()) {
+        return Some(metadata_rev.tags().to_vec());
+    }
+
+    None
+}
+
+fn url_encode(input: &str) -> String {
+    static ENCODE_REGEX: LazyLock<Regex> =
+        LazyLock::new(|| Regex::new(r"[#$&'()*+,/:;=?@\[\]\s]").unwrap());
+
+    ENCODE_REGEX
+        .replace_all(input, |caps: &Captures| match &caps[0] {
+            " " => "+".to_owned(),
+            _ => format!("%{:X}", &caps[0].as_bytes()[0]),
+        })
+        .into_owned()
+}
+
+fn format_uri_component(input: Option<String>) -> String {
+    input.as_deref().map(url_encode).unwrap_or("".to_owned())
+}
```

Comment:
Maybe us a crate for url encoding? Something like [percent_encoding](https://docs.rs/percent-encoding/latest/percent_encoding/index.html) seems to be the same which is used in `form_urlencoded`. Unless this is not fully url encoding and a simplified version? 

For example some special characters in the name of an artist or album might not be correctly converted. For example: https://open.spotify.com/track/2rMaTliflow3inMLEVYDVs

> also you can move functions that are only relevant inside another function into the function itself